### ISSUE TITLE
Display Part IDs and Property IDs

### ISSUE THREAD

**Issue by avidichard** [nptr/msstyleEditor#161]
When using GetThemeColor or any other uxtheme related API calls, they require IDs and NOT strings. I would like to suggest you to put the Part IDs in parenthesis beside their names. As for the Properties, I suggest you write their ID in the comment/note section you already have., I suggest it there to prevent information clogging for nothing but if someone clicks on the property, at least the information is available.

Please refer to this screenshot for better details.

![msstyleEditor Help](https://github.com/user-attachments/assets/759670a8-57b1-403c-91a5-d864c464bae0)

Property IDs can be found here: [https://docs.rs/windows-sys/latest/windows_sys/Win32/UI/Controls/constant.TMT_TEXTCOLOR.html](https://docs.rs/windows-sys/latest/windows_sys/Win32/UI/Controls/constant.TMT_TEXTCOLOR.html)

As for the Part IDs, the app seems to automatically sort them in the right order starting at 0 for the Common Properties. Info can be checked here for the known ones: [https://github.com/lxn/win/blob/master/uxtheme.go](https://github.com/lxn/win/blob/master/uxtheme.go) It gives a clue for others which are not documented such as DarkMode::ReadingPane which has BACKGROUNDCOLORS and LABEL Parts that are undocumented but use the IDs as listed, BACKGROUNDCOLORS being 1 and LABEL being 2. Tested with GetThemeColor API.

I know this is an editor but it happens to also be an excellent DEV tool that I discovered by surprise. The IDs would be very usefull!

**Comment by nptr** [comment]
Happy to hear its of good use!

- The class ID and part ID of the current selection is already diplayed in the status bar, in the lower left corner. Granted, its rather nondescript. I could write out "class" and "part" to make it more prominent.
![grafik](https://github.com/user-attachments/assets/33599247-dc7c-4b2b-aae7-a1d0183f40fb)

- The state ID is shown next to the state name as you seem to have noticed.
- The property ID is indeed nowhere visible. I will add it to the note section in the propery view as you suggested.

**Comment by avidichard** [comment]
I have not noticed that little area at all! That's indeed very helpfull to know. I still think it could be in the list to prevent the need of clicking on the part to get the part ID but at least it's there.

Also, when checking the Microsoft documentation on [SetWindowTheme](https://learn.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-setwindowtheme), They mention about the App Name and Sub ID list which they mention as CLSID names. The reason I am stating this is because when using DarkMode, SetWindowTheme uses "Explorer" as normal and "DarkMode_Explorer" for DarkMode. I came to believe that the app name was actually the string BEFORE the double colons "::" in the Class list.

![2024-11-29 13_11_36-CompatWindow](https://github.com/user-attachments/assets/f61e2964-ca30-4bf2-8371-31d8c53e7b7f)

I may be wrong but I have been studying theming an app for almost 4 weeks now delving into undocumented territories and this app revealed most of the missing information I needed but I do not know if maybe you have the proper knowledge of this to clarify these terms in the app and to know if I am right on guessing that since it works with SetWindowTheme.

I would have also loved to see your app use DarkMode since you brighten up the entire room when I open your app, I need to wear sunglasses LOL, kidding but Using Windows' actual theme would be neat.

And 1 last thing, your "Show Controls" also does not use or support DarkMode which I think it should or should be able to offer options to show default (Windows 7 style for Aero), DarkMode, LightMode and Themed (using the accent colors).

Thanks for your prompt reply and checking this out. Very nice app, lots of effort put into it and super usefull!

I'll look forward in an update with the property IDs, that will be super awesome and usefull!

**Comment by nptr** [comment]
I don't know with certainty and might be off at points. I believe `pszSubAppName` of `SetWindowTheme` refers to the name of the [Window Class](https://learn.microsoft.com/en-us/windows/win32/api/winuser/ns-winuser-wndclassexa) name `lpszClassName`. 

Every native control from the Windows common controls, registers such a name:
- Text boxes are "Edit" controls. Would be created via `CreateWindow()`, passing `Edit` as class name.
- List view control is `ListView`.
- Buttons, Groupboxes, Checkboxes etc are all actually just special cases of the generic "Button" class. See [this example](https://learn.microsoft.com/en-us/windows/win32/controls/create-a-button), where a push button (commonly known as button) is created with `CreateWindow()` and the `BS_DEFPUSHBUTTON` style.

Windows (like the `Explorer` window or msstyleEditor) also register such a window class. This class name corresponds to the class name in the visual style. Of course visual styles only consider the former.

When you create a control or window, it will have access to its own set of parts and properties from the visual style, based on its class name. Uxtheme supports inheritance of classes. A list view control, when used by the "Explorer" window (class) will get the special "Explorer::ListView" class instead of the regular ones.

`SetWindowTheme()` allows you to be treated as a different window class, or that you'd like to select your own set of classes that you (and your child controls?) like to use instead.

**Comment by nptr** [comment]
DarkMode is another beast. Microsoft chose to create dark versions of controls inside separate window classes "DarkMode_XXXX:YYYY". The problem is, these window classes are not automatically applied to all applications when dark mode is enabled. This is why msstyleEditor is bright as day (except when using a regular dark custom theme, not the Windows"dark" mode.)

An application must do [a lot of shenanigans](https://github.com/BlueMystical/Dark-Mode-Forms/blob/main/SourceFiles/DarkModeCS.cs) to get the Windows dark mode applied and even do custom rendering for some reason. 

It would have been much nicer had the dark mode been a regular, and separate theme (file). Or even a theme variant, which the themeing engines seems to support (see VMAP and VARIANT/<xxx> resources in the .msstyles file).

I believe this is for compatibility reasons. Lots of people probably hardcoded lots of colors in their apps instead of asking the themeing engine. This was somewhat fine because Windows always had a bright theme. Or maybe the reason Windows always had a bright theme (and did not have additional themes) was because of those applications and their unwillingness to just break stuff.

In any case, its a mess and MS has LOTS AND LOTS of baggage to carry. Doesn't help that they failed multiple times to establish a proper UI framework that would last more than a few years :(

**Comment by nptr** [comment]
That of course, makes your dark mode endeavours also quite hard. Is that what you are trying?

**Comment by avidichard** [comment]
Your ID's in v2.1.0.1 do not match. As I mentioned, TEXTCOLOR is ID 3803, you can actually se the whole list of IDs here in the left list (Ctrl+F will be your friend, the list is long). In your case, TEXTCOLOR is 204 AND FILLCOLOR is also 204 when it is actually 3802. I kind of guess that you are getting the ID from the TMT_COLOR and not TMT_TEXTCOLOR. I'd recheck those IDs.

And yes, I am indeed trying to use the Vanilla theming engine to apply Dark Theme to my app. And YES, it's a pain in the butt but I got arround finding some interesting shinanigans from within the msstyles file itself.

**Comment by nptr** [comment]
Oh damn. I am showing the type id instead of the property id :S
Serves me right doing things in a hurry,

**Comment by avidichard** [comment]
Seems alright to me now! Excellent job! I'd consider this as closed! Super Nice job and super handy. Still waiting for the "Class ID and Part ID to be in complete words instead of just "C" and "P" but that's minor. The important thing is that they are available once you know where to look.

I'll surely give some more insights on the DarkMode but here's a little something I found out while searching:

Windows Explorer's Title Bar colors is taken from:

**Class:** DarkMode::ReadingPane
**Background Part:** BACKGROUNDCOLORS
**Text Part:** LABEL

Yep! I have searched every possible "logical" parts and classes, the only color matching the title bar color in dark mode is THAT class. I use it to apply my color in my app. But again, changing the TITLE bar color is not the same and changing the window's background color and DWM manipulations do not do the job, actually, they don't even change a thing in the window, I'm building a Win32 app so things are more complicated than usual standard coding but I have came up with very good results, all I am missing is to have the ScroolBar backgrounds changed to their color since they stay white even with the proper Part applied to them and to get the control borders such as Textbox round corner and bottom border color as the accent color.
